### PR TITLE
Add PostgreSQL Auth Backend

### PR DESCRIPTION
**Pull request by rndevuk**
This commit enables the ability to use PostgreSQL as a Authentication backend.

### CODE REVIEW

**Comment by crazyR-MT** [comment]
I'm not a c++ programmer so if there is anything i can do to improve this then feel free to let me know below and i will do my best to accommodate.

I couldn't find a reference for database connection strings in the `world_format.txt` so i didn't add it into there.

The connection string is same format as other PostgreSQL  Features `pgsql_auth_connection = host=<host> user=<user> password=<password> dbname=<dbname>`

**Comment by MoNTE48** [comment]
Very well.
But also code for migration would be useful

**Comment by crazyR-MT** [comment]
migration is already in and didn't need amending.

File: src/database/database-postgresql.cpp [hunk 1]
```diff
@@ -631,4 +645,159 @@ void PlayerDatabasePostgreSQL::listPlayers(std::vector<std::string> &res)
...
+}
+
+void AuthDatabasePostgreSQL::createDatabase()
+{
+	if (getPGVersion() < 80400) {
+		createTableIfNotExists("auth",
+			"CREATE TABLE auth ("
+				"id SERIAL PRIMARY KEY,"
+				"name TEXT UNIQUE,"
+				"password VARCHAR(512),"
```
Comment (nerzhul) [inline src/database/database-postgresql.cpp]: please use TEXT not VARCHAR.

File: src/database/database-postgresql.cpp [hunk 2]
```diff
@@ -631,4 +645,159 @@ void PlayerDatabasePostgreSQL::listPlayers(std::vector<std::string> &res)
...
+				"privileges TEXT,"
```
Comment (nerzhul) [inline src/database/database-postgresql.cpp]: please use a dedicated proper table to store privileges

File: src/database/database-postgresql.cpp [hunk 3]
```diff
@@ -631,4 +645,159 @@ void PlayerDatabasePostgreSQL::listPlayers(std::vector<std::string> &res)
...
+				"last_login INT"
```
Comment (nerzhul) [inline src/database/database-postgresql.cpp]: instead of using a INT, use a timestamp

File: src/database/database-postgresql.cpp (hunk 1 above)
Comment (nerzhul) [reply src/database/database-postgresql.cpp]: also add NOT NULL constraint on required fields

**Comment by crazyR-MT** [comment]
Thanks for the feedback @nerzhul. i will start work on these changes over the weekend.

**Pull request closed without merging**